Add Tekton client plugin for GSOC 2021 (#4104)
jenkins-infra/jenkins.io: content/projects/gsoc/2021/project-ideas/tekton-client-plugin.adoc

Building on Kubernetes would mean that any CI/CD tool which has been Kubernetes-native built with building on Kubernetes as it’s main idea would thrive better than Jenkins itself. Leveraging this, we can use the capability of spinning pipelines on Tekton (Serverless CI/CD for Kubernetes) through Jenkins. Using the same vocabulary and ideas to execute Jobs from Jenkins and having the workloads of the Jobs run by Tekton in Kubernetes, this plugin aims to bridge the gap between Jenkins and Kubernetes.

Jenkins currently does not support Tekton, making it hard for users to use it without prior knowledge of the CI/CD. This plugin would enable users to carry their previous Jenkins knowledge to translate it to Tekton.

==== Project Details
This project idea proposes to implement a Jenkins plugin which allows users to trigger pipelines in Tekton.

=== Links

* link:https://github.com/fabric8io/kubernetes-client[Java Kubernetes Client]
* link:https://github.com/tektoncd/pipeline[Tekton Pipelines]
* https://github.com/jenkinsci/tekton-client-plugin[Tekton Client Plugin]

> === Quick-start
> 
> The project requires the student to start with plugin development from scratch and then work with understanding how to use the Cloudevents Java SDK to create and read events.
> 
> In the beginning, the student can work on understanding the Jenkins plugin development by writing code for creating a simple Build Step which creates a CloudEvent, and from there we can move to the GlobalPluginConfiguration on how Jenkins should listen to the cloudevents.

Reviewer: Per the [Google Doc](https://docs.google.com/document/d/1_mS0gLgM7oz7PanE0X_SkJ8iUo-BPVCixXMSDO7-VMk/edit#heading=h.vdp84117c6ct), the student should be able to start with the Tekton Client Plugin as a working proof of concept.

```suggestion
=== Quick-start

The project requires the student to start with plugin development from the [Tekton Client Plugin implemented as a proof of concept](https://github.com/jenkinsci/tekton-client-plugin).

Work with potential mentors and plugin maintainers to understand the plugin, identify its strengths and weaknesses, and better understand how to use Tekton from Jenkins.
```

---
layout: gsocprojectidea
title: "Tekton client plugin for Jenkins"
goal: "Trigger Tekton Pipelines from Jenkins Jobs."
category: Plugins
year: 2021
status: draft
sig: docs
skills:
- Java
- Kubernetes
- Tekton
mentors:
# To be confirmed

> #- "oleg_nenashev"

Reviewer 2: No, sorry. This year is quite rough in my case

---

=== Background
With the movement of traditional software architecture to microservices, there has been a growth in usage of container orchestration infrastructure (Kubernetes) and tooling around it. To adapt to this there have been plugins developed for Jenkins like the kubernetes-plugin which allow this transition to be smooth for the users to keep the same CI/CD platform and migrate their workloads. 

Building on Kubernetes would mean that any CI/CD tool which has been Kubernetes-native built with building on Kubernetes as it’s main idea would thrive better than Jenkins itself. Leveraging this, we can use the capability of spinning pipelines on Tekton (Serverless CI/CD for Kubernetes) through Jenkins. Using the same vocabulary and ideas to execute Jobs from Jenkins and having the workloads of the Jobs run by Tekton in Kubernetes, this plugin aims to bridge the gap between Jenkins and Kubernetes.

Jenkins currently does not support Tekton, making it hard for users to use it without prior knowledge of the CI/CD. This plugin would enable users to carry their previous Jenkins knowledge to translate it to Tekton.

==== Project Details

> This project idea proposes to implement a Jenkins plugin which allows users to trigger pipelines in Tekton.

Reviewer 2: Would be nice to expand it. Also, it is not clear why a new plugin is needed. Could https://github.com/jenkinsci/tekton-client-plugin just be extended (a few items from its roadmap)?

=== Links

> * link:https://github.com/fabric8io/kubernetes-client [Java Kubernetes Client]
> * link:https://github.com/tektoncd/pipeline [Tekton Pipelines]
> * https://github.com/jenkinsci/tekton-client-plugin [Tekton Client Plugin]

Reviewer 2: Broken rendering otherwise

```suggestion
* link:https://github.com/fabric8io/kubernetes-client[Java Kubernetes Client]
* link:https://github.com/tektoncd/pipeline[Tekton Pipelines]
* https://github.com/jenkinsci/tekton-client-plugin[Tekton Client Plugin]
```

=== Quick-start

The project requires the student to start with plugin development from scratch and then work with understanding how to use the Cloudevents Java SDK to create and read events.

> In the beginning, the student can work on understanding the Jenkins plugin development by writing code for creating a simple Build Step which creates a CloudEvent, and from there we can move to the GlobalPluginConfiguration on how Jenkins should listen to the cloudevents.

Reviewer 2: Looks like it was taken from another project

=== Newbie-friendly issues
link:https://issues.jenkins.io/browse/JENKINS-64666?jql=status%20%3D%20Open%20AND%20labels%20%3D%20newbie-friendly [Open newbie-friendly issues]

=== Skills to improve/study
* Java
* Docker,Kubernetes

> * Jenkins

Reviewer 2: ```suggestion
* Jenkins
* Tekton Pipelines
```

---
layout: gsocprojectidea
title: "Tekton client plugin for Jenkins"
goal: "Trigger Tekton Pipelines from Jenkins Jobs."
category: Plugins
year: 2021
status: draft
sig: docs
skills:
- Java
- Kubernetes
- Tekton
mentors:

> # To be confirmed
> #- "oleg_nenashev"

Reviewer: ```suggestion
```
